File: app/src/main/java/com/egecius/rxjava2_demo_2/CallbackFramework.java
package com.egecius.rxjava2_demo_2;


import android.os.Handler;
import android.util.Log;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

class CallbackFramework {

    private final ExecutorService executorService = Executors.newSingleThreadExecutor();
    private final Handler mHandler = new Handler();

    interface Callback {
        void onSuccess();
    }

    void returnOnMainThread(Callback callback) {
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                callback.onSuccess();
            }
        }, 300);
    }

    void returnOnBackgroundThread(Callback callback) {
        Log.v("Eg:CallbackFramework:30", "returnOnBackgroundThread");

        executorService.execute(new Runnable() {
            @Override
            public void run() {
                callback.onSuccess();
            }
        });
    }
}

File: app/src/main/java/com/egecius/rxjava2_demo_2/CallbackFramework.kt
package com.egecius.rxjava2_demo_2


import android.os.Handler
import android.util.Log
import java.util.concurrent.Executors

internal class CallbackFramework {

    private val executorService = Executors.newSingleThreadExecutor()
    private val mHandler = Handler()

    internal interface Callback {
        fun onSuccess()
    }

    fun returnOnMainThread(callback: Callback) {
        mHandler.postDelayed({ callback.onSuccess() }, 300)
    }

    fun returnOnBackgroundThread(callback: Callback) {
        Log.v("Eg:CallbackFramework:30", "returnOnBackgroundThread")

        executorService.execute { callback.onSuccess() }
    }
}
